python/lsst/daf/butler/_butler.py

```diff
@@ -525,6 +528,83 @@ def get_known_repos(cls) -> set[str]:
         """
         return ButlerRepoIndex.get_known_repos()
 
+    @classmethod
+    def parse_dataset_uri(cls, uri: str) -> tuple[str, DatasetId]:
+        """Extract the butler label and dataset ID from a dataset URI.
+
+        Parameters
+        ----------
+        uri : `str`
+            The dataset URI to parse.
+
+        Returns
+        -------
+        label : `str`
+            The label associated with the butler repository from which this
+            dataset originates.
+        dataset_id : `DatasetId`
+            The ID of the dataset.
+
+        Notes
+        -----
+        Supports dataset URIs of the forms ``ivo://rubin/butler_label/UUID``
+        and ``butler://butler_label/UUID``. In ``ivo`` URIs the butler label
+        can include ``/`` and the leading ``/`` is always stripped. If the
+        repository label starts with ``/`` then it must be doubled up. e.g.,
+
+            ivo://rubin//repo/main/82d79caa-0823-4300-9874-67b737367ee0
+
+        will return a label of ``/repo/main``.
+
+        This method does not attempt to check that the dataset exists in the
+        labeled butler.
+        """
+        parsed = urllib.parse.urlparse(uri)
+        if parsed.scheme == "ivo":
+            # TODO: Validate netloc component.
+            label, id_ = os.path.split(parsed.path)
+            # Strip the leading /.
+            label = label[1:]
+        elif parsed.scheme == "butler":
+            label = parsed.netloc
+            # Need to strip the leading /.
+            id_ = parsed.path[1:]
+        else:
+            raise ValueError(f"Unrecognized URI scheme: {uri!r}")
+        if not label:
+            raise ValueError(f"No butler repository label found in uri {uri!r}")
+        try:
+            dataset_id = uuid.UUID(hex=id_)
+        except Exception as e:
+            e.add_note(f"Error extracting dataset ID from uri {uri!r} with dataset ID string {id_!r}")
+            raise
+
+        return label, dataset_id
+
+    @classmethod
+    def get_dataset_from_uri(cls, uri: str) -> DatasetRef | None:
```

Comment:
I like your idea of a config object that could be set up by the caller and passed through.

One other idea would be a dependency injection pattern.  Maybe declare a protocol like
```
class ButlerFactoryProtocol(Protocol):
    create(label: str) -> Butler

# (or just `Callable[[str], Butler]
```
which `get_dataset_from_uri` could take as a parameter.  Then `LabeledButlerFactory` could grow a method `bind(access_token: str) -> ButlerFactoryProtocol` to create a factory with the access token already bound in.  That would let callers do whatever they want to customize the instance creation.

Reply:
At the moment the access token is in fact shared since it comes from Gafaelfawr and just represents the user's identity.  Though no guarantees it will stay they way if we end up needing to delegate authentication between IDF and USDF, or some other unfortunate thing.

Reply:
Trying to think what you said above. For:

```python
my_factory = factory.bind(access_token=token) 
ref = Butler.get_dataset_from_uri(dataset_uri, my_factory)
```
doesn't this mean that the access_token is required to be the same for every butler label?

Don't we need:

```python
my_factory = factory.bind("dp02": {"access_token": token}, "dp02-direct": {"access_token": None})
```
?

Reply:
There is only one access token per HTTP request -- Gafaelfawr generates it to identify the user.  It is not associated with the repository in any way.  So at least at the moment there's only one access token.  If you pass an access token to `LabeledButlerFactory` for a DirectButler label it just ignores it.

The advantage to the dependency injection pattern is that this code wouldn't have to know anything at all about authentication or where Butlers come from -- it makes that the caller's responsibility.  This code just knows that if it has a label, the caller will work out how to get it the matching Butler.  So if things get more complicated later, this interface doesn't have to change.  (One example might be caching the Butler if someone wants to call `get_dataset_from_uri` multiple times in the same request -- though that might be better solved by wrapping this function in a separate object.)

Your config object idea would work well too, allowing us to add more options later if `LabeledButlerFactory` gets more complicated.  It's just a little less flexible and it couples this function tightly to `LabeledButlerFactory`.